### DISCUSSION TITLE
Release 1.1.0

### DISCUSSION THREAD

**Discussion by TheR1D** [TheR1D/shell_gpt#431] [Announcements]
https://github.com/TheR1D/shell_gpt/assets/16740832/721ddb19-97e7-428f-a0ee-107d027ddd59

# OpenAI Library
ShellGPT has now integrated the OpenAI Python library for handling API requests. This integration simplifies the development and maintenance of the ShellGPT code base. Additionally, it enhances user experience by providing more user-friendly error messages, complete with descriptions and potential solutions.

# Function Calling
[Function calls](https://platform.openai.com/docs/guides/function-calling) is a powerful feature OpenAI provides. It allows LLM to execute functions in your system, which can be used to accomplish a variety of tasks. ShellGPT has a convenient [way to define functions](https://github.com/TheR1D/shell_gpt#function-calling) and use them. To install [default functions](https://github.com/TheR1D/shell_gpt/tree/main/sgpt/default_functions/) run: 
```shell
sgpt --install-functions
```
This will add a function for LLM to execute shell commands and to execute Apple Scripts (on macOS). More details in demo video and README.md.

# Options
* Shortcut option `-c` for `—code`.
* Shortcut option `-lc` for `--list-chats`
* Shortcut option `-lr` for `--list-roles`
* New `—functions` option, enables/disable function calling.
* New `—install-functions` option, installs default functions.

# Config
* New config variable `OPENAI_FUNCTIONS_PATH`
* New config variable `OPENAI_USE_FUNCTIONS`
* New config variable `SHOW_FUNCTIONS_OUTPUT`

# Minor Changes
* Code optimisation
* Cache optimisations for function calls

<hr /><em>This discussion was created from the release <a href='https://github.com/TheR1D/shell_gpt/releases/tag/1.1.0'>1.1.0</a>.</em>

**Comment by hcal** [comment]
I think I misunderstood your original post describing how functions would work.  I probably had something else in mind and projected my idea on your words.  what you've done here is even cooler than what I imagined,  also I really appreciated the production value on the video.  Great job.  


I use this tool everyday, if you're interested in any support via patreon, GitHub, etc... I'm sure I'm not the only one willing to pitch in.  Either way, amazing work.

**Reply by TheR1D** [reply]
@hcal thank you for your positive feedback! If you're interested in supporting my work further, you can do so via [GitHub sponsors](https://github.com/sponsors/TheR1D). Either way, I'm just happy to know that you're finding value in the tool. Thanks again!
